src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+-- TODO: something like this instead:
+--    type Tag = T.Text
+--    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
+type RawName = D.Name
+
+-- invariant: the raw name in `Name s` is contained in the list in the `ScopeVal s`
+newtype Name (n::S) = UnsafeMakeName RawName
+                      deriving (Show, Eq, Ord)
+
+newtype Env (n::S) (a:: *) = UnsafeMakeEnv (LM.LazyMap RawName a)
+                             deriving (Functor, Foldable, Traversable)
+type Scope n = Env n ()
+
+emptyEnv :: Env (n:=>:n) a
+emptyEnv = UnsafeMakeEnv mempty
+
+voidScopeEnv :: Env VoidScope a
+voidScopeEnv = UnsafeMakeEnv mempty
+
+infixl 1 <.>
+(<.>) :: Env (n1:=>:n2) a -> Env (n2:=>:n3) a -> Env (n1:=>:n3) a
+(<.>) (UnsafeMakeEnv m1) (UnsafeMakeEnv m2) = UnsafeMakeEnv (m2 <> m1)
+
+infixl 1 <>>
+(<>>) :: Env n a -> Env (n:=>:l) a -> Env l a
+(<>>) (UnsafeMakeEnv m1) (UnsafeMakeEnv m2) = UnsafeMakeEnv (m2 <> m1)
+
+envLookup :: Env n a -> Name n -> a
+envLookup (UnsafeMakeEnv m) (UnsafeMakeName name) =
+  case LM.lookup name m of
+    Just x -> x
+    Nothing -> error "Env lookup should never fail"
+
+envAsScope :: Env n a -> Scope n
+envAsScope (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.asUnitLazyMap m
+
+projectName :: Scope (n:=>:l) -> Name l -> Either (Name n) (Name (n:=>:l))
+projectName (UnsafeMakeEnv m) (UnsafeMakeName name) =
+  case LM.lookup name m of
+    Nothing -> Left  $ UnsafeMakeName name
+    Just () -> Right $ UnsafeMakeName name
+
+injectNameL :: FreshExt n l -> Name n -> Name l
+injectNameL _ (UnsafeMakeName name) = UnsafeMakeName name
+
+injectNameR :: Name (n:=>:l) -> Name l
+injectNameR (UnsafeMakeName name) = UnsafeMakeName name
+
+data PlainBinder (n::S) (l::S) where
+  UnsafeMakeBinder :: RawName -> PlainBinder n l
+  Ignore           :: PlainBinder n n
+
+type E = S -> *       -- expression-y things, covariant in the S param
+type B = S -> S -> *  -- binder-y things, covariant in the first param and
+                      -- contravariant in the second
+
+-- A `FreshExt n l` means that `l` extends `n` with only fresh names
+data FreshExt (n::S) (l::S) = UnsafeMakeExt
+
+emptyFreshExt :: FreshExt n n
+emptyFreshExt = UnsafeMakeExt
+
+composeFreshExt :: FreshExt n h -> FreshExt h l -> FreshExt n l
+composeFreshExt _ _ = UnsafeMakeExt
+
+data NameTraversal (e::E) (m:: * -> *) (i::S) (o::S) where
+  NameTraversal :: (Name i -> m (e o))
+                -> RenameEnv (i:=>:i') o
+                -> NameTraversal e m i' o
+
+type RenameTraversal = NameTraversal Name
+
+-- traverses free names, possibly renaming bound ones as needed
+class HasNamesE (e::E) where
+  traverseNamesE :: Monad m => Scope o -> RenameTraversal m i o -> e i -> m (e o)
+
+-- A fresh binder, hiding its new output scope and exposing a renaming
+-- env that can give you those names. Used mainly for the result of
+-- `traverseNamesB` and similar.
+data FreshBinder (b::B) (n::S) (i::S) where
+  FreshBinder :: FreshExt o o' -> b o o' -> RenameEnv i o'
```

Comment:
Yes, renaming the variables is confusing, sorry. This one is a head scratcher and I don't love using it. But it seems like the most general thing for the result of `TraverseNamesB`. It might be easier to switch to the rank-2 continuation-oriented version of existentials for `TraverseNamesB`.